File: 0013-roman-to-integer/0013-roman-to-integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        arr = {}
        arr['I'] = 1
        arr['V'] = 5
        arr['X'] = 10
        arr['L'] = 50
        arr['C'] = 100
        arr['D'] = 500
        arr['M'] = 1000
        sums = 0
        i = len(s)-1
        while i>=0:
            if i>0 and (s[i-1]=='I') and (s[i] == 'V' or s[i] == 'X'):
                sums+=arr[s[i]]-arr[s[i-1]]
                i-=2
            elif i>0 and (s[i-1]=='X') and (s[i] == 'L' or s[i] == 'C'):
                sums+=arr[s[i]]-arr[s[i-1]]
                i-=2
            elif i>0 and (s[i-1]=='C') and (s[i] == 'D' or s[i] == 'M'):
                sums+=arr[s[i]]-arr[s[i-1]]
                i-=2
            else:
                sums+=arr[s[i]]
                i-=1
        return sums
        

File: 0013-roman-to-integer/0013-roman-to-integer.java
class Solution {
    public int romanToInt(String s) {
        int n = s.length();
        int num = 0;
        Map<Character,Integer> map = new HashMap<>();
        map.put('I',1);
        map.put('V',5);
        map.put('X',10);
        map.put('L',50);
        map.put('C',100);
        map.put('D',500);
        map.put('M',1000);
        int i=0;
        while(i<n){
            char ch = s.charAt(i);
            char next = (i + 1 < n) ? s.charAt(i + 1) : ' ';
            
            if((ch=='I') && (next=='V' || next=='X')){
                num+=map.get(next)-map.get(ch);
                i+=2;
            }
            else if((ch=='X') && (next=='L' || next=='C')){
                num+=map.get(next)-map.get(ch);
                i+=2;
            }
            else if((ch=='C') && (next=='D' || next=='M')){
                num+=map.get(next)-map.get(ch);
                i+=2;
            }
            else{
                num+=map.get(ch);
                i++;
            }
        }
        return num;
    }
}